\documentclass[10pt]{article}
\usepackage[margin=1in]{geometry}
\usepackage{longtable}
\usepackage[pdfborder={0 0 0},hyperfootnotes=false]{hyperref}
\usepackage[title]{appendix}

\newcommand{\mailtourl}[1]{\href{mailto:#1}{\tt #1}}
\newcommand{\tagvalue}[1]{\tt #1}
\newcommand{\tagregex}[1]{\tt #1}

\begin{document}

\input{SAMstrict.ver}
\title{SAM Strict Specification}
\author{Daniel L Cameron}
\date{\headdate}
\maketitle
\begin{quote}\small
The master version of this document can be found at
\url{https://github.com/samtools/hts-specs}.\\
This printing is version~\commitdesc\ from that repository,
last modified on the date shown above.
\end{quote}
\vspace*{1em}

\noindent
This document is a companion to the {\sl Sequence Alignment/Map Format
Specification} that defines the SAM file format.\footnote{See
\href{http://samtools.github.io/hts-specs/SAMv1.pdf}{\tt SAMv1.pdf} at \url{https://github.com/samtools/hts-specs}.}
The SAM file format defines the syntax required for a file to be
a valid SAM file. It does not require such files to be semantically
valid and internally consistent.
This document describes a set of additional semantic restrictions
for which the subset of syntactically valid SAM files that comply
with these restrictions can be described as \textit{SAM strict
compliant}.

\renewcommand{\abstractname}{Introduction}
\begin{abstract}

The SAM specifications have been instrumental in standardising
the file formats used for sequence alignment. A large ecosystem of
bioinformatics tools is now capable of reading and/or writing
SAM files. Unfortunately, many tools that read SAM files are tightly
coupled to a particular upstream tool 
and fail to correctly execute on valid SAM files written by other
tools. In part, this is due to the lack of semantic restrictions
inherent in the SAM file format. A syntactically valid SAM file
can be both internally inconsistent and semantically nonsensical.

The purpose of this document is to provide a baseline of semantic
validity for which tools should comply with when outputing SAM
files, and tools which input SAM files can safely assume when
they require input files to be \textit{SAM strict compliant}.

\end{abstract}

\section{Headers}

\subsection{HD}

\paragraph{}

The @HD line must be present, with either the SO tag or the GO tag (but not both) specified.

\paragraph{}

As the SAM specifications does not explicitly define the queryname sort order, SO=queryname
should be avoided in favour of GO=queryname.

\subsection{SQ}

\paragraph{}

@SQ SN fields must be unique. Multiple headers with the same SN must not be present.

\paragraph{}

Every SAM record with an non-* RNAME must have a corresponding @SQ header with matching SN.

\section{General}

\subsection{Ordering}

\paragraph{}

If a SO or GO @HD header tag is defined, the order of records must be consistent with this ordering.

\subsection{CIGAR}

\paragraph{}

All CIGAR strings must have at least one CIGAR operator.

\paragraph{}

All CIGAR operators must have a non-zero positive length

\paragraph{}

All adjacent CIGAR operators must be different.

\paragraph{}

TODO: should we disallow I/D operators at the ends of reads? There was some ambiguity as to how deletions interacted with POS but I think the spec has been clarified in favour of the BWA interpretation since that discussion.

\subsection{Flags}

TODO:
- # reads in template is correct
- ...

\subsection{Alignments}

\paragraph{}

Each read must have exactly one non-supplementary non-secondary record.

\paragraph{}

No supplementary or secondary alignments may exist for reads with an
unmapped primary (non-supplementary non-secondary record) alignment.

\paragraph{}

All read alignments should have CIGARs consistent with the 

\paragraph{}

All read alignments must have CIGARs with matching read length. That is,
the sum of lengths of the M/I/S/=/X operations must be equal for all mapped read alignment.
This means that chimeric alignments must include either soft clipping or
hard clipping CIGAR operations for read bases which were not aligned.

\paragraph{}

All read alignments with SEQ not equal to * must have a SEQ consistent
with all other read alignments. That is, chimeric and secondary alignments that have non-* SEQ
must be consistent with all other records defining a segment sequence. For
example, 10th base in a read cannot be be A in one alignment, but T in another.

\paragraph{}

All read alignments with QUAL not equal to * must have a SEQ consistent
with all other read alignments.

\subsection{Mate alignments}

\paragraph{}

Unmapped reads must have RNAME and POS identical to that of the primary
non-supplementary alignment of first mapped read from the originating template.
For paired-end or mate-pair sequencing, this equivalent
to setting the RNAME/POS of unmapped reads to the RNAME/POS of the mate.

\paragraph{}

For templates with multiple reads, RNEXT and PNEXT must match the
alignment of the non-supplementary primary alignment of the next read.
As per the SAM specifications, for the last read, the next read is the first read in the template.

\paragraph{}

If all segments in a template are unmapped, RNAME must be set as `*' and POS as 0.

\subsection{Reference bounds}

\paragraph{}

Read alignments must not extend past the start or end of the aligned RNAME.

\subsection{Mapping Qualities}

\paragraph{}

All mapping quality scores, including those defined in tags must be within the range [0, 255].
A value 255 indicates that the mapping quality is not available and must only be used if the
mapping quality field is required. For example, a mapping quality field value is required for
MAPQ field and the mapq portion of the SA tag, but as the AM is optional, a mapping quality
field value is not required and the AM tag should be omitted entirely if a mapping quality is
not available.

\section{SAM Tags}

\subsection{Standard Tags}

\paragraph{}

No record can include any reserved tags not defined in the
{\sl Sequence Alignment/Map Optional Fields Specification}.
Non-standard tags must start X, Y, Z or a lowercase letter as per the SAM specifications.

\paragraph{}

The \textit{type} of all \textit{standard tags} must match the type
defined in the {\sl Sequence Alignment/Map Optional Fields Specification}.

\paragraph{}

All tag values must be consistent with the format
defined in the {\sl Sequence Alignment/Map Optional Fields Specification}.

\subsection{SA}

For the purpose of this section, a \textit{SA record set} is a set of SAM records
from a single \textit{read} which collectively represent a single \textit{chimeric alignment}.

\paragraph{}

All SA tag values must satisfy the SA tag regular expression
defined in the {\sl Sequence Alignment/Map Optional Fields Specification}.

\paragraph{}

All records referenced in the SA tag of a given record must have a SA tag defined.

\paragraph{}

All records referenced in the SA tag of a given record must include the given
record in their SA tag.

\paragraph{}

All records in a \textit{SA record set} with a FI tag defined must have the same FI tag value.

TODO: which other tags?

\paragraph{}

All records referenced in the SA tag must exist with matching \textit{rname, pos, strand, CIGAR}.

\paragraph{}

The SA \textit{mapq} of all references to a given record in a \textit{SA record set} must
match the record mapping quality.

\paragraph{}

The SA \textit{NM} of all references to a given record in a \textit{SA record set} must
match the record \textit{NM} tag value.

\paragraph{}

All records except 1 in a \textit{SA record set} must have the 0x800 (supplementary alignment) FLAG bit set.

\paragraph{}

The first SA record in all supplementary alignment records must be the canonical non-supplementary alignment.

\paragraph{}

All records in a \textit{SA record set} must have the same 0x100 (secondary alignment) FLAG value.

\paragraph{}

All records in a \textit{SA record set} must have FLAG bit 0x4 (segment unmapped) not set.

\paragraph{}

All records in a \textit{SA record set} must align at least one read base that does not
overlap with any other alignments in the \textit{SA record set}.
That is, a chimeric alignment cannot contain superfluous alignment records.

\end{document}
